addons/resource/models/resource.py

```diff
@@ -987,6 +986,129 @@ def _get_unavailable_intervals(self, start, end):
+                start = tz_dates[(tz, start_dt)]
+            else:
+                start = start_dt.astimezone(tz)
+                tz_dates[(tz, start_dt)] = start
+            if (tz, end_dt) in tz_dates:
+                end = tz_dates[(tz, end_dt)]
+            else:
+                end = end_dt.astimezone(tz)
+                tz_dates[(tz, end_dt)] = end
+
+            personal_leaves = personal_leaves_intervals[res.id] if personal_leaves_intervals.get(res.id) else False
+            global_leaves = global_leave_per_calendar[res.calendar_id.id] if global_leave_per_calendar.get(res.calendar_id.id) else False
+            resource_work_intervals = res._leaves_interval_gantt_view(start, end, personal_leaves, global_leaves, dict_attendance_per_calendar[res.calendar_id][res.id])
+
+            work_intervals = [(start, stop) for start, stop, meta in resource_work_intervals]
+            # start + flatten(intervals) + end
+            work_intervals = [start_datetime] + list(chain.from_iterable(work_intervals)) + [end_datetime]
+            # put it back to UTC
+            work_intervals = list(map(lambda dt: dt.astimezone(utc), work_intervals))
+            # pick groups of two
+            work_intervals = list(zip(work_intervals[0::2], work_intervals[1::2]))
+            leaves_mapping[res.id] = work_intervals
+        return leaves_mapping
+
+    def _personal_leaves_intervals(self, start_dt, end_dt):
+        """
+        :return: a dict { res.id : list[date_start, date_stop] } with the personal leaves of each resource
+        """
+        domain = [
+            ('time_type', '=', 'leave'),
+            ('resource_id', 'in', self.ids),
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        leaves = self.env['resource.calendar.leaves'].search_read(domain, fields=['date_from', 'date_to', 'resource_id'])
+        dict_personal_leaves = {}
+        for leave in leaves:
+            if dict_personal_leaves.get(leave['resource_id'][0]):
+                dict_personal_leaves[leave['resource_id'][0]].append([leave['date_from'], leave['date_to']])
+            else:
+                dict_personal_leaves[leave['resource_id'][0]] = [[leave['date_from'], leave['date_to']]]
+        return dict_personal_leaves
+
+    def _global_leaves_per_calendar(self, start_dt, end_dt, calendars_set):
+        """
+        :return: a dict { res.id : list[date_start, date_stop] } with the global leaves of each calendar. Each dict entry also contains the global leaves that have
+        no calendar_id
+        """
+        domain = [
+            ('resource_id', '=', False),
+            ('time_type', '=', 'leave'),
+            ('calendar_id', 'in', [False] + calendars_set.ids),
+            ('date_from', '<=', end_dt),
+            ('date_to', '>=', start_dt),
+        ]
+        leaves = self.env['resource.calendar.leaves'].search_read(domain, fields=['date_from', 'date_to', 'calendar_id'])
```

Comment:
Can't we group the request with the previous one? Records will be easy to distinguish base on their resource_id. :thinking:

Reply:
No that's the core of the problem. If we take the following case : one leave A in calendar B, and the user U with calendar C.
We need both the leave A and the global leave of calendar C. If we merge the two domain, we will excludes the calendar B from the search.
If you really want to merge the domain and do only one search, we'll have to fetch all the leave of the db with something like 
```
domain = [
            ('time_type', '=', 'leave'),
            ('date_from', '<=', end_dt),
            ('date_to', '>=', start_dt),
        ]
```
which does not seems like a good idea.

Reply:
Clarified in mp